src/server/pachyderm_test.go

```diff
@@ -3641,47 +3641,81 @@ func TestCreateDatum(t *testing.T) {
-		return nil
+	testTimeToFirstDatum := func(input *pps.Input) {
+		var eg errgroup.Group
+		var listDatumTimeToFirstDatum, createDatumTimeToFirstDatum float64
+		eg.Go(func() error {
+			ctx, cf := context.WithCancel(c.Ctx())
+			defer cf()
+			req := &pps.ListDatumRequest{Input: input}
+			start := time.Now()
+			client, err := c.PpsAPIClient.ListDatum(ctx, req)
+			if err != nil {
+				return err
+			}
+			_, err = client.Recv()
+			if err != nil {
+				return err
+			}
+			listDatumTimeToFirstDatum = time.Since(start).Seconds()
+			return nil
+		})
+		eg.Go(func() error {
+			ctx, cf := context.WithCancel(c.Ctx())
+			defer cf()
+			client, err := c.PpsAPIClient.CreateDatum(ctx)
+			if err != nil {
+				return err
+			}
+			req := &pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input}}}
+			start := time.Now()
+			if err := client.Send(req); err != nil {
+				return err
+			}
+			_, err = client.Recv()
+			if err != nil {
+				return err
+			}
+			createDatumTimeToFirstDatum = time.Since(start).Seconds()
+			return nil
+		})
+		require.NoError(t, eg.Wait())
+		t.Logf("listDatumTimeToFirstDatum: %f seconds", listDatumTimeToFirstDatum)
+		t.Logf("createDatumTimeToFirstDatum: %f seconds", createDatumTimeToFirstDatum)
+		require.True(t, createDatumTimeToFirstDatum < listDatumTimeToFirstDatum)
+	}
+
+	t.Run("PFSInput", func(t *testing.T) {
+		t.Parallel()
+		input := client.NewPFSInput(pfs.DefaultProjectName, repo, "/*")
+		testTimeToFirstDatum(input)
 	})
-	eg.Go(func() error {
-		ctx, cf := context.WithCancel(c.Ctx())
-		defer cf()
-		client, err := c.PpsAPIClient.CreateDatum(ctx)
-		if err != nil {
-			return err
-		}
-		req := &pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input}}}
-		start := time.Now()
-		if err := client.Send(req); err != nil {
-			return err
-		}
-		_, err = client.Recv()
-		if err != nil {
-			return err
-		}
-		createDatumTimeToFirstDatum = time.Since(start).Seconds()
-		return nil
+	t.Run("UnionInput", func(t *testing.T) {
+		t.Parallel()
+		input := client.NewUnionInput(
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/*"),
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/*"),
+		)
+		testTimeToFirstDatum(input)
+	})
+	t.Run("CrossInput", func(t *testing.T) {
+		t.Parallel()
+		input := client.NewCrossInput(
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/file-??"),
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/file-???"),
+		)
+		testTimeToFirstDatum(input)
+	})
+	t.Run("JoinInput", func(t *testing.T) {
+		t.Parallel()
+		input := client.NewJoinInput(
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file-?*(??)0", "$1", "", false, false, nil),
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file-?0(??)0", "$1", "", false, false, nil),
+		)
+		testTimeToFirstDatum(input)
 	})
```

Comment:
This should be a table-driven test:

```go
testData := []struct{name string; input *pfs.Input}{
  {"PFSInput", client.NewPFSInput(...)},
  {"UnionInput", client.NewUnionInput(...)},
  ...
}
for _, test := range testData {
   t.Run(test.name, func(t *testing.T) { 
      t.Parallel()
      var eg errgroup.Group
      var listDatumTimeToFirstDatum, createDatumTimeToFirstDatum float64
      ...
    })
}
```
Also, I am not sure how reliable it is to collect one timing data point each time the test is run.  Nothing guarantees that each goroutine is given the same scheduling latency or the same amount of time on the CPU, so even if CreateDatum is faster than ListDatum, you aren't guaranteed to measure that every time.  It's just a recipe for a flaky test.  This could be a benchmark (`func BenchmarkCreateDatum(b *testing.B)` if you want to make sure that you collect enough timing samples to be confident in the result.

Reply:
I changed it to a benchmark to avoid flakiness. Running the benchmark a few times confirmed the improved latency in getting the first datum.